File: src/lug_ffi.erl
-module(lug_ffi).
-export([drop_byte/1, compile_binary_pattern/1, split_before/2]).

%% borrowed from https://github.com/DanielleMaywood/glexer
drop_byte(String) ->
    case String of
        <<_, Rest/bytes>> -> Rest;
        _ -> String
    end.

%% borrowed from https://github.com/lpil/splitter
compile_binary_pattern(Pattern) ->
    binary:compile_pattern(Pattern).

%% borrowed from https://github.com/lpil/splitter
split_before(Pattern, String) ->
    case binary:match(String, Pattern) of
        % No delimiter found
        nomatch ->
            {String, <<"">>};
        {Index, _Length} ->
            {binary:part(String, 0, Index), binary:part(String, Index, byte_size(String) - Index)}
    end.

File: src/lug_ffi.mjs
// borrowed from https://github.com/DanielleMaywood/glexer
export function slice_bytes(string, start, size) {
  return string.slice(start, start + size);
}

// borrowed from https://github.com/DanielleMaywood/glexer
export function drop_byte(string) {
  return string.slice(1);
}

// borrowed from https://github.com/lpil/splitter
export function compile_binary_pattern(patterns) {
  let pattern = "";
  let cursor = patterns;
  while (cursor.tail) {
    if (pattern !== "") pattern += "|";
    pattern += escapeRegExp(cursor.head);
    cursor = cursor.tail;
  }
  return new RegExp(pattern);
}

// borrowed from https://github.com/lpil/splitter
export function split_before(splitter, string) {
  const match = string.match(splitter);

  if (!match) return [string, ""]; // No delimiter found

  const split_point = match.index;
  return [string.slice(0, split_point), string.slice(split_point)];
}
